### ISSUE TITLE
Imprecision with duration

### ISSUE THREAD

**Issue by hb9eik** [mlco2/codecarbon#292]
* CodeCarbon version: 2.0.0a3
* Python version: 3.8.10
* Operating System: 20.04.4 LTS (Focal Fossa)

### Description
Trying to verify data as par bug #286 I found that the timing reported in csv are different from the timing extracted in debug

### What I Did
I added some debug to verify data, I attach the emission_tracker.py modified and two datasheet with problem evidenced

first run (with emission_tracker
[emissions.ods](https://github.com/mlco2/codecarbon/files/8312090/emissions.ods)
[emissions1.ods](https://github.com/mlco2/codecarbon/files/8312091/emissions1.ods)
[example.zip](https://github.com/mlco2/codecarbon/files/8312105/example.zip)
[example.zip](https://github.com/mlco2/codecarbon/files/8312109/example.zip)
[example.zip](https://github.com/mlco2/codecarbon/files/8312111/example.zip)
.py modified to log duration and power
root@aurora:/home/giovanni# python3
Python 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from codecarbon import EmissionsTracker
>>> tracker = EmissionsTracker()
[codecarbon INFO @ 19:42:05] [setup] RAM Tracking...
[codecarbon INFO @ 19:42:05] [setup] GPU Tracking...
[codecarbon INFO @ 19:42:05] No GPU found.
[codecarbon INFO @ 19:42:05] [setup] CPU Tracking...
[codecarbon INFO @ 19:42:05] Tracking Intel CPU via RAPL interface
[codecarbon INFO @ 19:42:06] >>> Tracker's metadata:
[codecarbon INFO @ 19:42:06]   Platform system: Linux-5.4.0-104-generic-x86_64-with-glibc2.29
[codecarbon INFO @ 19:42:06]   Python version: 3.8.10
[codecarbon INFO @ 19:42:06]   Available RAM : 7.660 GB
[codecarbon INFO @ 19:42:06]   CPU count: 4
[codecarbon INFO @ 19:42:06]   CPU model: Intel(R) Core(TM) i5-6400 CPU @ 2.70GHz
[codecarbon INFO @ 19:42:06]   GPU count: None
[codecarbon INFO @ 19:42:06]   GPU model: None
>>> tracker.start()
>>> tracker.stop()
[codecarbon INFO @ 19:42:16] last duration
[codecarbon INFO @ 19:42:16] 4.231994152069092
[codecarbon INFO @ 19:42:16] power 
[codecarbon INFO @ 19:42:16] Power(kW=0.002872594356536865)
[codecarbon INFO @ 19:42:16] energy
[codecarbon INFO @ 19:42:16] Energy(kWh=3.3768895883696356e-06)
[codecarbon INFO @ 19:42:16] Energy consumed for RAM : 0.000003 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 19:42:16] last duration
[codecarbon INFO @ 19:42:16] 4.231994152069092
[codecarbon INFO @ 19:42:16] power 
[codecarbon INFO @ 19:42:16] Power(kW=0.007106655628418483)
[codecarbon INFO @ 19:42:16] energy
[codecarbon INFO @ 19:42:16] Energy(kWh=8.354256961176643e-06)
[codecarbon INFO @ 19:42:16] Energy consumed for all CPUs : 0.000008 kWh. All CPUs Power : 7.1066556284184825 W
[codecarbon INFO @ 19:42:16] 0.000012 kWh of electricity used since the begining.
2.4870030685038107e-06
>>> tracker.stop()
[codecarbon INFO @ 19:42:26] last duration
[codecarbon INFO @ 19:42:26] 9.235263586044312
[codecarbon INFO @ 19:42:26] power 
[codecarbon INFO @ 19:42:26] Power(kW=0.002872594356536865)
[codecarbon INFO @ 19:42:26] energy
[codecarbon INFO @ 19:42:26] Energy(kWh=7.369212794000362e-06)
[codecarbon INFO @ 19:42:26] Energy consumed for RAM : 0.000011 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 19:42:26] last duration
[codecarbon INFO @ 19:42:26] 9.235263586044312
[codecarbon INFO @ 19:42:26] power 
[codecarbon INFO @ 19:42:26] Power(kW=0.0029233203087876175)
[codecarbon INFO @ 19:42:26] energy
[codecarbon INFO @ 19:42:26] Energy(kWh=7.4993426661361384e-06)
[codecarbon INFO @ 19:42:26] Energy consumed for all CPUs : 0.000016 kWh. All CPUs Power : 2.9233203087876176 W
[codecarbon INFO @ 19:42:26] 0.000027 kWh of electricity used since the begining.
5.639136826052749e-06
>>> tracker.stop()
[codecarbon INFO @ 19:42:31] last duration
[codecarbon INFO @ 19:42:31] 5.500642776489258
[codecarbon INFO @ 19:42:31] power 
[codecarbon INFO @ 19:42:31] Power(kW=0.002872594356536865)
[codecarbon INFO @ 19:42:31] energy
[codecarbon INFO @ 19:42:31] Energy(kWh=4.389198721407865e-06)
[codecarbon INFO @ 19:42:31] Energy consumed for RAM : 0.000015 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 19:42:31] last duration
[codecarbon INFO @ 19:42:31] 5.500642776489258
[codecarbon INFO @ 19:42:31] power 
[codecarbon INFO @ 19:42:31] Power(kW=0.0029997661856951867)
[codecarbon INFO @ 19:42:31] energy
[codecarbon INFO @ 19:42:31] Energy(kWh=4.5835117223613775e-06)
[codecarbon INFO @ 19:42:31] Energy consumed for all CPUs : 0.000020 kWh. All CPUs Power : 2.999766185695187 W
[codecarbon INFO @ 19:42:31] 0.000036 kWh of electricity used since the begining.
7.541351440131829e-06
>>> 

Second run (with emission_tracker.py modified to not use too many times time() function)

Python 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from codecarbon import EmissionsTracker
>>> tracker = EmissionsTracker()
[codecarbon INFO @ 20:29:21] [setup] RAM Tracking...
[codecarbon INFO @ 20:29:21] [setup] GPU Tracking...
[codecarbon INFO @ 20:29:21] No GPU found.
[codecarbon INFO @ 20:29:21] [setup] CPU Tracking...
[codecarbon INFO @ 20:29:21] Tracking Intel CPU via RAPL interface
[codecarbon INFO @ 20:29:22] >>> Tracker's metadata:
[codecarbon INFO @ 20:29:22]   Platform system: Linux-5.4.0-104-generic-x86_64-with-glibc2.29
[codecarbon INFO @ 20:29:22]   Python version: 3.8.10
[codecarbon INFO @ 20:29:22]   Available RAM : 7.660 GB
[codecarbon INFO @ 20:29:22]   CPU count: 4
[codecarbon INFO @ 20:29:22]   CPU model: Intel(R) Core(TM) i5-6400 CPU @ 2.70GHz
[codecarbon INFO @ 20:29:22]   GPU count: None
[codecarbon INFO @ 20:29:22]   GPU model: None
>>> tracker.start()
>>> tracker.stop()
[codecarbon INFO @ 20:29:33] last duration
[codecarbon INFO @ 20:29:33] 4.96051025390625
[codecarbon INFO @ 20:29:33] power 
[codecarbon INFO @ 20:29:33] Power(kW=0.002872594356536865)
[codecarbon INFO @ 20:29:33] energy
[codecarbon INFO @ 20:29:33] Energy(kWh=3.958203822476207e-06)
[codecarbon INFO @ 20:29:33] Energy consumed for RAM : 0.000004 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 20:29:33] last duration
[codecarbon INFO @ 20:29:33] 4.96051025390625
[codecarbon INFO @ 20:29:33] power 
[codecarbon INFO @ 20:29:33] Power(kW=0.006186939242413069)
[codecarbon INFO @ 20:29:33] energy
[codecarbon INFO @ 20:29:33] Energy(kWh=8.525104320079163e-06)
[codecarbon INFO @ 20:29:33] Energy consumed for all CPUs : 0.000009 kWh. All CPUs Power : 6.186939242413069 W
[codecarbon INFO @ 20:29:33] 0.000012 kWh of electricity used since the begining.
2.6464613262217385e-06
>>> tracker.stop()
[codecarbon INFO @ 20:29:39] last duration
[codecarbon INFO @ 20:29:39] 6.687378168106079
[codecarbon INFO @ 20:29:39] power 
[codecarbon INFO @ 20:29:39] Power(kW=0.002872594356536865)
[codecarbon INFO @ 20:29:39] energy
[codecarbon INFO @ 20:29:39] Energy(kWh=5.336145773813712e-06)
[codecarbon INFO @ 20:29:39] Energy consumed for RAM : 0.000009 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 20:29:39] last duration
[codecarbon INFO @ 20:29:39] 6.687378168106079
[codecarbon INFO @ 20:29:39] power 
[codecarbon INFO @ 20:29:39] Power(kW=0.002835847277477796)
[codecarbon INFO @ 20:29:39] energy
[codecarbon INFO @ 20:29:39] Energy(kWh=5.267884214302243e-06)
[codecarbon INFO @ 20:29:39] Energy consumed for all CPUs : 0.000014 kWh. All CPUs Power : 2.8358472774777956 W
[codecarbon INFO @ 20:29:39] 0.000023 kWh of electricity used since the begining.
4.8945156837023205e-06
>>> tracker.stop()
[codecarbon INFO @ 20:29:45] last duration
[codecarbon INFO @ 20:29:45] 6.191872835159302
[codecarbon INFO @ 20:29:45] power 
[codecarbon INFO @ 20:29:45] Power(kW=0.002872594356536865)
[codecarbon INFO @ 20:29:45] energy
[codecarbon INFO @ 20:29:45] Energy(kWh=4.9407608229645915e-06)
[codecarbon INFO @ 20:29:45] Energy consumed for RAM : 0.000014 kWh. RAM Power : 2.8725943565368652 W
[codecarbon INFO @ 20:29:45] last duration
[codecarbon INFO @ 20:29:45] 6.191872835159302
[codecarbon INFO @ 20:29:45] power 
[codecarbon INFO @ 20:29:45] Power(kW=0.002953912672162692)
[codecarbon INFO @ 20:29:45] energy
[codecarbon INFO @ 20:29:45] Energy(kWh=5.080625453388055e-06)
[codecarbon INFO @ 20:29:45] Energy consumed for all CPUs : 0.000019 kWh. All CPUs Power : 2.953912672162692 W
[codecarbon INFO @ 20:29:45] 0.000033 kWh of electricity used since the begining.
7.019049574289082e-06
>>> 

```

**Comment by benoit-cty** [comment]
Hello,
Sorry for the delay !
With @h-sdl we think we have found a solution : https://github.com/mlco2/codecarbon/pull/293

To put CodeCarbon in debug mode, could you edit/create a file `.codecarbon.config` and put these line in it:
```
[codecarbon]
log_level = DEBUG
```
So you will see lines like:
`[codecarbon DEBUG @ 21:57:20][] GPU : 206.78 W during 29.98 s [measurement time: 0.0151]`

We think that the delta in time you see is due to the fact that we lost the "measurement time" in last_duration.

**Comment by hb9eik** [comment]
Which version can I install to test the solution ?
I'm really willing to test this solution.
Thanks again for all your work

**Comment by benoit-cty** [comment]
You're welcome, thanks for reporting this.
It's not yet published to Pypi, due to dependencies problems affecting many Python projects.
To install it in a specific environment you could do:
```
git clone https://github.com/mlco2/codecarbon.git
cd codecarbon
python3 -m venv .venv
source .venv/bin/activate
pip install -e .
```

**Comment by benoit-cty** [comment]
I made a test:
Duration in CSV : 108.567s
Sum of last_duration : 8.55504+9.9964+9.9976+9.9958+9.99843+9.9962+9.9984+9.9967+9.9994+9.99418+10.0043=108.53245
The goal of last_duration is not to keep absolute time duration, only the duration between two measures.
And we indeed have a problem because we have only one last_duration and made many measure (CPU, GPU, RAM).
Some of them could last up to a second. So it will be better to have a last_duration inside each hardware class for more accuracy.
But with all the other drawbacks of CodeCarbon (we do not know the consumption of whole computer, nor the PUE of the datacenter) I don't know if it is worth it.
